lib/internal/events/event_emitter.js

```diff
@@ -0,0 +1,707 @@
+        }
+      }
+    }
+  };
+
+// If you're updating this function definition, please also update any
+// re-definitions, such as the one in the Domain module (lib/domain.js).
+EventEmitter.init = function(opts) {
+  let thisPrototype;
+
+  if (this[kImpl] === undefined || (thisPrototype = ObjectGetPrototypeOf(this))[kImpl] === this[kImpl]) {
+    this[kImpl] = new FastEventEmitter(this);
+    this[kIsFastPath] = true;
+  }
+
+  const thisEvents = this._events;
+  const impl = this[kImpl];
+
+  if (thisEvents === undefined ||
+      thisEvents === (thisPrototype || ObjectGetPrototypeOf(this))._events) {
+    // In fast path we don't want to have __proto__: null as it will cause the object to be in dictionary mode
+    // and will slow down the access to the properties by a lot (around 2x)
+    impl._events = this[kIsFastPath] ? {} : { __proto__: null };
+    impl._eventsCount = 0;
+    impl[kShapeMode] = false;
+  } else if (this[kInitialEvents] !== undefined) {
+    impl._events = this[kInitialEvents];
+    this[kInitialEvents] = undefined;
+    impl[kShapeMode] = true;
+  }
+
+  this._maxListeners = this._maxListeners || undefined;
+
+
+  if (opts?.captureRejections) {
+    validateBoolean(opts.captureRejections, 'options.captureRejections');
+    this[kCapture] = Boolean(opts.captureRejections);
+  } else {
+    // Assigning the kCapture property directly saves an expensive
+    // prototype lookup in a very sensitive hot path.
+    this[kCapture] = EventEmitter.prototype[kCapture];
+  }
+};
+
+EventEmitter.prototype[kSwitchToSlowPath] = function() {
+  if (this[kIsFastPath] === false) {
+    return;
+  }
+
+  if (this[kIsFastPath] === true) {
+    this[kIsFastPath] = false;
+    this[kImpl] = SlowEventEmitter.fromFastEventEmitter(this[kImpl]);
+    return;
+  }
+
+  this[kImpl] = new SlowEventEmitter(this);
+  this[kIsFastPath] = false;
+
+};
+
+/**
+ * Increases the max listeners of the event emitter.
+ * @param {number} n
+ * @returns {EventEmitter}
+ */
+EventEmitter.prototype.setMaxListeners = function setMaxListeners(n) {
+  validateNumber(n, 'setMaxListeners', 0);
+  this._maxListeners = n;
+  return this;
+};
+
+function _getMaxListeners(that) {
+  if (that._maxListeners === undefined)
+    return EventEmitter.defaultMaxListeners;
+  return that._maxListeners;
+}
+/**
+ * Returns the current max listener value for the event emitter.
+ * @returns {number}
+ */
+EventEmitter.prototype.getMaxListeners = function getMaxListeners() {
+  return _getMaxListeners(this);
+};
+
+
+/**
+ * Synchronously calls each of the listeners registered
+ * for the event.
+ * @param {string | symbol} type
+ * @param {...any} [args]
+ * @returns {boolean}
+ */
+EventEmitter.prototype.emit = function emit(type, ...args) {
+  // Users can call emit in the constructor before even calling super causing this[kImpl] to be undefined
+  const impl = this[kImpl];
+
+  // The order here is important as impl === undefined is slower than type === 'error'
```

Comment:
I have no idea why checking `type === 'error'` first is faster when running `ee-emit.js` benchmark and slower is `impl === undefined` first